File: 3207-maximum-points-after-enemy-battles/3207-maximum-points-after-enemy-battles.java
class Solution {
    public long maximumPoints(int[] enemyEnergies, int currentEnergy) {
        Arrays.sort(enemyEnergies);

        if(currentEnergy == 0 || enemyEnergies[0] > currentEnergy) return 0;
        long points = 0;
        int i = 0;
        int j = enemyEnergies.length-1;

        while(i <= j){
            if(currentEnergy >= enemyEnergies[i]){
                points += currentEnergy / enemyEnergies[i];
                currentEnergy = currentEnergy %  enemyEnergies[i];
            }
            else{
                currentEnergy += enemyEnergies[j];
                j--;
            }
        }    

        return points; 
    }
}

File: 3207-maximum-points-after-enemy-battles/3207-maximum-points-after-enemy-battles.py
class Solution:
    def maximumPoints(self, enemyEnergies: List[int], currentEnergy: int) -> int:
        enemyEnergies.sort()
        points = 0
        i, j = 0, len(enemyEnergies) - 1
        
       
        while i <= j:
            if currentEnergy >= enemyEnergies[i]:
                points += (currentEnergy // enemyEnergies[i])
                currentEnergy = (currentEnergy % enemyEnergies[i])
               
            elif points > 0:
                currentEnergy += enemyEnergies[j]
                j -= 1
            else:
                break
                
        return points